pkg/chains/storage/ca/ca.go

```diff
@@ -0,0 +1,350 @@
+/*
+Copyright 2020 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ca
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	attestationpb "github.com/grafeas/grafeas/proto/v1beta1/attestation_go_proto"
+	commonpb "github.com/grafeas/grafeas/proto/v1beta1/common_go_proto"
+	pb "github.com/grafeas/grafeas/proto/v1beta1/grafeas_go_proto"
+	"github.com/pkg/errors"
+	"github.com/tektoncd/chains/pkg/artifacts"
+	"github.com/tektoncd/chains/pkg/chains/formats"
+	"github.com/tektoncd/chains/pkg/config"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/oauth"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+)
+
+const (
+	StorageBackendCA  = "containeranalysis"
+	ProjectNameFormat = "projects/%s"
+	NoteNameFormat    = "projects/%s/notes/%s"
+)
+
+// Backend is a storage backend that stores signed payloads in the TaskRun metadata as an annotation.
+// It is stored as base64 encoded JSON.
+type Backend struct {
+	logger         *zap.SugaredLogger
+	tr             *v1beta1.TaskRun
+	client         pb.GrafeasV1Beta1Client
+	cfg            config.Config
+	occurrenceRefs []string // store occurrence IDs that are automatically generated during the time of creation
+}
+
+func NewStorageBackend(logger *zap.SugaredLogger, tr *v1beta1.TaskRun, cfg config.Config) (*Backend, error) {
+	// add TypeMeta information to taskrun objects
+	addTypeInformationToObject(tr)
+
+	// build connection through grpc
+	// implicit uses Application Default Credentials to authenticate.
+	// Requires `gcloud auth application-default login` to work locally
+	ctx := context.Background()
+	creds, err := oauth.NewApplicationDefault(ctx, "https://www.googleapis.com/auth/cloud-platform")
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := grpc.Dial("dns:///containeranalysis.googleapis.com",
+		grpc.WithTransportCredentials(credentials.NewTLS(&tls.Config{})),
+		grpc.WithDefaultCallOptions(grpc.PerRPCCredentials(creds)),
+	)
+	if err != nil {
+		return nil, err
+	}
+	// defer conn.Close()
```

Comment:
is this needed?

Reply:
Should I delete this `conn.Close()` for now, and add that later if there is a big change in storage that allows us to close all connections?